libc/stdlib_alloc.c

```diff
@@ -0,0 +1,318 @@
+
+	void *ptr2;
+	ptr2 = malloc(sizeof(char));
+	TEST_ASSERT_NOT_NULL(ptr2);
+	TEST_ASSERT_TRUE((uintptr_t)ptr2 % sizeof(char) == 0);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(char), malloc_usable_size(ptr2));
+#endif
+	memset(ptr2, 0x28, sizeof(char));
+	TEST_ASSERT_EACH_EQUAL_HEX8(0x28, ptr2, sizeof(char));
+
+	free(ptr0);
+	free(ptr1);
+	free(ptr2);
+}
+
+
+TEST(stdlib_malloc, malloc_zero)
+{
+	void *ptr;
+	ptr = malloc(0);
+#ifdef phoenix
+	TEST_ASSERT_NULL(ptr);
+#else
+	TEST_ASSERT_NOT_NULL(ptr);
+#endif
+	free(ptr);
+}
+
+TEST(stdlib_malloc, calloc_basic)
+{
+	int *ptr = calloc(BLOCK_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(BLOCK_SIZE, malloc_usable_size(ptr));
+#endif
+	free(ptr);
+}
+
+TEST(stdlib_malloc, calloc_zero)
+{
+	int *ptr0 = calloc(0, sizeof(int));
+#ifdef phoenix
+	TEST_ASSERT_NULL(ptr0);
+#else
+	TEST_ASSERT_NOT_NULL(ptr0);
+#endif
+
+	int *ptr1 = calloc(BLOCK_SIZE, 0);
+#ifdef phoenix
+	TEST_ASSERT_NULL(ptr0);
+#else
+	TEST_ASSERT_NOT_NULL(ptr0);
+#endif
+
+	free(ptr0);
+	free(ptr1);
+}
+
+TEST(stdlib_malloc, calloc_one)
+{
+	int *ptr = calloc(1, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, *ptr);
+
+	free(ptr);
+}
+
+TEST(stdlib_malloc, calloc_large)
+{
+	int *ptr = calloc(BLOCK_SIZE * LARGE_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE * LARGE_SIZE);
+	free(ptr);
+}
+
+TEST(stdlib_malloc, realloc_null)
+{
+	int *ptr = realloc(NULL, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+#ifndef phoenix
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(sizeof(int), malloc_usable_size(ptr));
+#endif
+	free(ptr);
+}
+
+TEST(stdlib_malloc, realloc_zero_size)
+{
+	int *ptr = calloc(1, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+
+	int *ptr1 = realloc(ptr, 0);
+	TEST_ASSERT_NULL(ptr1);
+}
+
+TEST(stdlib_malloc, realloc_resize)
+{
+	int NUMBERS[BLOCK_SIZE];
+	int *ptr = (int *)calloc(BLOCK_SIZE, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EACH_EQUAL_INT(0, ptr, BLOCK_SIZE);
+	for (int i = 0; i < BLOCK_SIZE; i++) {
+		ptr[i] = i;
+		NUMBERS[i] = i;
+	}
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr, BLOCK_SIZE);
+	TEST_ASSERT_TRUE((uintptr_t)ptr % sizeof(int) == 0);
+
+
+	int *ptr1 = (int *)realloc(ptr, BLOCK_SIZE * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_EQUAL_INT_ARRAY(NUMBERS, ptr1, BLOCK_SIZE);
+	TEST_ASSERT_TRUE((uintptr_t)ptr1 % sizeof(int) == 0);
+
+	free(ptr1);
+}
+
+TEST(stdlib_malloc, realloc_resize_small)
+{
+	int *ptr = (int *)calloc(1, 2 * sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr);
+	TEST_ASSERT_EQUAL_INT(0, ptr[0]);
+	ptr[0] = 1;
+
+	int *ptr1 = (int *)realloc(ptr, sizeof(int));
+	TEST_ASSERT_NOT_NULL(ptr1);
+	TEST_ASSERT_EQUAL_INT(1, ptr1[0]);
+	TEST_ASSERT_TRUE((uintptr_t)ptr1 % sizeof(int) == 0);
+
+	free(ptr1);
+}
+
+TEST(stdlib_malloc, realloc_resize_large)
```

Comment:
This test do the same thing as `realloc_resize_small` but uses larger block sizes. Although, both of them shrinks allocated memory. I would like to see a test that is similar to this but memory is increased -> e.g. malloc(BLOCK_SIZE), realloc(ptr, BLOCK_SIZE*2).

To sum up:
I suggest having two test cases: `realloc_resize_smaller` and `realloc_resize_larger`.
In the first one we are resizing block from larger to smaller (e.g two cases: `2*sizeof(int)` -> `sizeof(int)` and `2*(big chunk of memory)` -> `(big chunk of memory)`.
The second one is similar but from smaller to larger.